### ISSUE TITLE
set_optimizer_state_dict does not support partial state dict

### ISSUE THREAD

**Issue by ppwwyyxx** [pytorch/pytorch#132628]
### 🐛 Describe the bug

```python
import torch
from torch import nn
from torch.distributed.checkpoint.state_dict import set_optimizer_state_dict, get_optimizer_state_dict
mod = nn.Sequential(nn.Linear(3, 3), nn.Linear(3, 3))
opt = torch.optim.SGD(mod.parameters(), momentum=0.9)

mod(torch.randn(1, 3)).sum().backward()
opt.step()

sd = get_optimizer_state_dict(mod, opt)
print(sd)

sd['state'].pop('0.weight')

set_optimizer_state_dict(mod, opt, sd)  # fail
# opt.load_state_dict(sd)  # succeed
sd = get_optimizer_state_dict(mod, opt)
print(sd)
```

The error is :
```
Traceback (most recent call last):
  File "/HOME/a.py", line 18, in <module>
    set_optimizer_state_dict(mod, opt, sd)  # fail
  File "/HOME/.pyenv/versions/torch24/lib/python3.10/site-packages/torch/distributed/checkpoint/state_dict.py", line 1223, in set_optimizer_state_dict
    _load_optim_state_dict(model, optimizers, optim_state_dict, info)
  File "/HOME/.pyenv/versions/torch24/lib/python3.10/site-packages/torch/distributed/checkpoint/state_dict.py", line 867, in _load_optim_state_dict
    optim_state_dict = _split_optim_state_dict(
  File "/HOME/.pyenv/versions/torch24/lib/python3.10/site-packages/torch/distributed/checkpoint/state_dict.py", line 834, in _split_optim_state_dict
    state[fqn] = cast(DictValueType, optim_state_dict[_STATE])[fqn]
KeyError: '0.weight'
```

This behavior is different from `optimizer.load_state_dict`.
This pattern is useful when fine-tuning a model with more learnable parameters - in this case some parameters have existing states to load while some don't.

### Versions

torch2.4

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @LucasLLC @MeetVadakkanchery @mhorowitz @pradeepfn

**Comment by fegin** [comment]
How do you handle the extra learnable parameters? Are you using `strict` option or are you using a separate checkpoint?  We can add `strict` option support for `set_optimizer_state_dict` but you can also use a separate checkpoint for the extra learnable parameters.

**Comment by ppwwyyxx** [comment]
We use the `strict` option for model state dict. Note that in this particular use case, the extra parameters do not exist before fine-tuning and are expected to initialize randomly, so it does not make sense to load a separate checkpoint.

**Comment by YurongYou** [comment]
Same here. Wondering if there are any updates on this? Thanks!

**Comment by fegin** [comment]
@ppwwyyxx, @YurongYou The regular Pytorch optimizer will also raise an exception if parameter group does not match: https://fburl.com/ok1us2to. So I'm not sure how this case has to be supported in general. If one parameter doesn't exist before fine-tuning, then we should use a different optimizer for it. Otherwise, the parameter group won't match.

**Comment by ppwwyyxx** [comment]
You're right that the regular optimizer does not support this either. For this case we actually have implemented our own "strict=False" logic to match a partial state dict's param group against a new optimizer, but it might not make sense for pytorch to support that feature in general.

**Comment by fegin** [comment]
It is difficult to support this feature without the support from PyTorch optimizer. The main concern is that PyTorch optimizer doesn't support dynamically changing the parameters being optimized. In such a case, I would still argue that the best way is to have a separate optimizer or users should have a proprietary implementation to fit their own requirements.

@janeyx99 do you have any comment about this feature? 

I'll close the issue later unless there are other proposal to make this feature move forward.

**Comment by janeyx99** [comment]
I'd expect this to be called by the user whenever convenient:
(1) if the state_dict to be loaded has the state for the extra parameters, then one should add_param_group before loading the state
(2) if the state_dict to be loaded does NOT have the extra state, then calling load_state_dict should be first, with add_param_group after to initiate with empty state.

**Comment by fegin** [comment]
@ppwwyyxx I think the second condition @janeyx99 mentioned should work for your use case.

**Comment by YurongYou** [comment]
The situation I encounter is that the `set_model_state_dict` cannot even load partial ckpt 
```python
from torch import nn
from torch.distributed.checkpoint.state_dict import set_model_state_dict, StateDictOptions
mod = nn.Sequential(nn.Linear(3, 3), nn.Linear(3, 3))

model_state_dict = mod.state_dict()
model_state_dict.pop('0.weight')
options = StateDictOptions(
    broadcast_from_rank0=True,
    full_state_dict=True,
    cpu_offload=True,
    strict=False,
)
set_model_state_dict(mod, model_state_dict, options=options) # fail
```

**Comment by YurongYou** [comment]
The situation I encountered is that `set_model_state_dict` cannot load partial ckpts on DDP models, even with `strict=False` option
```python
from torch import nn
import torch.distributed as dist
from torch.nn.parallel import DistributedDataParallel as DDP
from torch.distributed.checkpoint.state_dict import set_model_state_dict, StateDictOptions
dist.init_process_group(world_size=1, rank=0, backend='gloo')

mod = DDP(nn.Sequential(nn.Linear(3, 3), nn.Linear(3, 3)))

model_state_dict = mod.state_dict()
model_state_dict.pop('module.0.weight')
options = StateDictOptions(
    strict=False,
)
set_model_state_dict(mod, model_state_dict, options=options) # fail
```

**Comment by YurongYou** [comment]
Can we reopen this issue? I think the issue has not been resolved yet

**Comment by fegin** [comment]
@YurongYou It's fixed with https://github.com/pytorch/pytorch/pull/143038

**Comment by YurongYou** [comment]
> @YurongYou It's fixed with #143038

Nice! Thank you! Will this be in torch 2.6?

**Comment by fegin** [comment]
@YurongYou Unfortunately, no. But we can see if we can make it to 2.6.1
